arch/arm64/core/mmu.c

```diff
@@ -1351,3 +1360,197 @@ void z_arm64_swap_mem_domains(struct k_thread *incoming)
+	 * doesn't end up being zero as this would be seen as a free entry.
+	 */
+	desc |= PTE_BLOCK_DESC_AP_RO;
+
+	*pte = desc;
+	MMU_DEBUG("page_out: virt=%#lx location=%#lx\n", virt, location);
+	debug_show_pte(pte, XLAT_LAST_LEVEL);
+
+	sync_domains(virt, CONFIG_MMU_PAGE_SIZE, "page_out");
+	invalidate_tlb_page(virt);
+}
+
+void arch_mem_page_in(void *addr, uintptr_t phys)
+{
+	uintptr_t virt = (uintptr_t)addr;
+	uint64_t *pte = get_pte_location(&kernel_ptables, virt);
+	uint64_t desc;
+
+	__ASSERT((phys & ~PTE_PHYSADDR_MASK) == 0, "");
+
+	__ASSERT(pte != NULL, "");
+	desc = *pte;
+	__ASSERT(!is_free_desc(desc), "");
+
+	/* mark the entry valid again to the hardware */
+	desc &= ~PTE_DESC_TYPE_MASK;
+	desc |= PTE_PAGE_DESC;
+
+	/* store the physical address */
+	desc &= ~PTE_PHYSADDR_MASK;
+	desc |= phys;
+
+	/* mark as clean */
+	desc |= PTE_BLOCK_DESC_AP_RO;
+
+	/* and make it initially unaccessible to track unaccessed pages */
+	desc &= ~PTE_BLOCK_DESC_AF;
+
+	*pte = desc;
+	MMU_DEBUG("page_in: virt=%#lx phys=%#lx\n", virt, phys);
+	debug_show_pte(pte, XLAT_LAST_LEVEL);
+
+	sync_domains(virt, CONFIG_MMU_PAGE_SIZE, "page_in");
+	invalidate_tlb_page(virt);
+}
+
+enum arch_page_location arch_page_location_get(void *addr, uintptr_t *location)
+{
+	uintptr_t virt = (uintptr_t)addr;
+	uint64_t *pte = get_pte_location(&kernel_ptables, virt);
+	uint64_t desc;
+	enum arch_page_location status;
+
+	if (!pte) {
+		return ARCH_PAGE_LOCATION_BAD;
+	}
+	desc = *pte;
+	if (is_free_desc(desc)) {
+		return ARCH_PAGE_LOCATION_BAD;
+	}
+
+	switch (desc & PTE_DESC_TYPE_MASK) {
+	case PTE_PAGE_DESC:
+		status = ARCH_PAGE_LOCATION_PAGED_IN;
+		break;
+	case PTE_INVALID_DESC:
+		status = ARCH_PAGE_LOCATION_PAGED_OUT;
+		break;
+	default:
+		return ARCH_PAGE_LOCATION_BAD;
+	}
+
+	*location = desc & PTE_PHYSADDR_MASK;
+	return status;
+}
+
+uintptr_t arch_page_info_get(void *addr, uintptr_t *phys, bool clear_accessed)
+{
+	uintptr_t virt = (uintptr_t)addr;
+	uint64_t *pte = get_pte_location(&kernel_ptables, virt);
+	uint64_t desc;
+	uintptr_t status = 0;
+
+	if (!pte) {
+		return ARCH_DATA_PAGE_NOT_MAPPED;
+	}
+	desc = *pte;
+	if (is_free_desc(desc)) {
+		return ARCH_DATA_PAGE_NOT_MAPPED;
+	}
+
+	switch (desc & PTE_DESC_TYPE_MASK) {
+	case PTE_PAGE_DESC:
+		status |= ARCH_DATA_PAGE_LOADED;
+		break;
+	case PTE_INVALID_DESC:
+		/* page not loaded */
+		break;
+	default:
+		return ARCH_DATA_PAGE_NOT_MAPPED;
+	}
+
+	if (phys) {
+		*phys = desc & PTE_PHYSADDR_MASK;
+	}
+
+	if ((status & ARCH_DATA_PAGE_LOADED) == 0) {
+		return status;
+	}
+
+	if ((desc & PTE_BLOCK_DESC_AF) != 0) {
+		status |= ARCH_DATA_PAGE_ACCESSED;
+	}
+
+	if ((desc & PTE_BLOCK_DESC_AP_RO) == 0) {
+		status |= ARCH_DATA_PAGE_DIRTY;
+	}
+
+	if (clear_accessed) {
+		desc &= ~PTE_BLOCK_DESC_AF;
+		*pte = desc;
+		MMU_DEBUG("page_info: virt=%#lx (clearing AF)\n", virt);
+		debug_show_pte(pte, XLAT_LAST_LEVEL);
+		sync_domains(virt, CONFIG_MMU_PAGE_SIZE, "unaccessible");
+		invalidate_tlb_page(virt);
+	}
```

Comment:
Hm... `clear_accessed` is supposed to only clear the access state (actually the `ARCH_DATA_PAGE_ACCESSED` state, and associated hardware bits) and not about making next access to generate any exceptions. If NRU is used, there is no need to invalidate the TLB but only needs the access bits cleared.

So maybe guard the TLB invalidation with #ifdef?